### PR TITLE
Adding support for RVFI on cv32e40p env

### PR DESCRIPTION
**Pull request by YoannPruvost**
First step toward supporting RVFI based test, done by merging the original env and testbench with the cv32e40x one.
For now, only the ci tests are supported

### CODE REVIEW

**Comment by MikeOpenHWGroup** [comment]
Hi @YoannPruvost.  Is this PR related to [cv32e40p #717](https://github.com/openhwgroup/cv32e40p/pull/717)?  Should that PR be completed before this one?

**Comment by YoannPruvost** [comment]
> Hi @YoannPruvost. Is this PR related to [cv32e40p #717](https://github.com/openhwgroup/cv32e40p/pull/717)? Should that PR be completed before this one?

Hi @MikeOpenHWGroup yes, the pull request on the core should be completed first. This update of the env requires the RVFI to be implemented on the cv32e40p.

**Comment by MikeOpenHWGroup** [comment]
Thanks @YoannPruvost.  I added the DO NOT MERGE label. We can remove it when [cv32e40p #717](https://github.com/openhwgroup/cv32e40p/pull/717) has been merged.  This PR will also need to update [cv32e40p/sim/ExternalRepos.mk](https://github.com/openhwgroup/core-v-verif/blob/9a26828d802bd77ec445a7c441e7e8234671ef20/cv32e40p/sim/ExternalRepos.mk#L18) to point to the new hash of the cv32e40p repo.

File: cv32e40p/sim/ExternalRepos.mk
Comment (MikeOpenHWGroup) [inline cv32e40p/sim/ExternalRepos.mk:18]: This branch will not compile with `dsim`, see [cv32e40p #719](https://github.com/openhwgroup/cv32e40p/pull/719).

**Review by MikeOpenHWGroup — CHANGES_REQUESTED** [review]
Wow!  Impressive work @YoannPruvost.  You did a great job of learning the core-v-verif structure and updating the CV32E40P environment to match the E40X environment.  I have made requests for several changes:
1. Several of the compile-guards reference CV32E40X, not CV32E40P.
2. The core_cntrl_cfg needs an update to support debug testing (see my comments).
3. There is a bug in the RVFI tracer that prevents the core from compiling with dsim, see cv32e40p pull-request [#719](https://github.com/openhwgroup/cv32e40p/pull/719)

File: cv32e40p/env/uvme/uvma_cv32e40p_core_cntrl_agent.sv
Comment (YoannPruvost) [reply cv32e40p/env/uvme/uvma_cv32e40p_core_cntrl_agent.sv]: done

File: cv32e40p/env/uvme/uvma_cv32e40p_core_cntrl_agent.sv
Comment (YoannPruvost) [reply cv32e40p/env/uvme/uvma_cv32e40p_core_cntrl_agent.sv]: done

File: cv32e40p/env/uvme/uvma_cv32e40p_core_cntrl_cntxt.sv
Comment (YoannPruvost) [reply cv32e40p/env/uvme/uvma_cv32e40p_core_cntrl_cntxt.sv]: done

File: cv32e40p/env/uvme/uvme_cv32e40p_cfg.sv
Comment (YoannPruvost) [reply cv32e40p/env/uvme/uvme_cv32e40p_cfg.sv:157]: done

File: cv32e40p/env/uvme/uvme_cv32e40p_env.sv
Comment (YoannPruvost) [reply cv32e40p/env/uvme/uvme_cv32e40p_env.sv]: done

File: cv32e40p/env/uvme/uvme_cv32e40p_env.sv
Comment (YoannPruvost) [reply cv32e40p/env/uvme/uvme_cv32e40p_env.sv]: done

File: cv32e40p/env/uvme/vseq/uvme_cv32e40p_core_cntrl_base_seq.sv
Comment (YoannPruvost) [reply cv32e40p/env/uvme/vseq/uvme_cv32e40p_core_cntrl_base_seq.sv]: done

File: cv32e40p/env/uvme/vseq/uvme_cv32e40p_core_cntrl_fetch_toggle_seq.sv
Comment (YoannPruvost) [reply cv32e40p/env/uvme/vseq/uvme_cv32e40p_core_cntrl_fetch_toggle_seq.sv]: done

**Comment by MikeOpenHWGroup** [comment]
Thanks for the update @YoannPruvost - much appreciated.  I'll approve and merge this once  [cv32e40p #719](https://github.com/openhwgroup/cv32e40p/pull/719) is merged in.

**Comment by MikeOpenHWGroup** [comment]
@pascalgouedo, I have (finally) determined why I could not get this PR to work.  There is an issue with [cv32e40p_rvfi.sv](https://github.com/openhwgroup/cv32e40p/blob/bddbd38fb0b5ced6208090691002b7ee021dcc8d/bhv/cv32e40p_rvfi.sv#L687) in which a set of members of `class insn_trace_t` are written to with non-blocking assignments.  It is not legal SystemVerilog to use non-blocking assignments for automatic variables, and by default all object members in SV are automatic.  It turns out that Cadence Xcelium allows this, but Metrics Dsim does not. (I believe Dsim's interpretation of the SV LRM is correct, so these should be compile-time errors under Xcelium as well.)

My first attempt to resolve the problem was to remove all non-blocking assignments (`<=`) with blocking assignments (`=`).   This resolved the compile issue with Dsim but introduced functional errors in simulation.  My second attempt was to declare all members of `class insn_trace_t` to be static variables.  This should be safe because as far as I can tell, the objects constructed from this class are "newed" once and only once during a simulation.

I believe that merging this PR will work for Cadence Xcelium, but it will kill our CI flow which runs under Dsim, so I would like to discuss the updates I think are necessary for both the cv32e40p and core-v-verif repos to resolve this.  You should have a meeting invitation in your Inbox already.

**Pull request closed without merging**
